internal/controllers/cryostat_controller_test.go

```diff
@@ -831,6 +836,67 @@ var _ = Describe("CryostatController", func() {
 				t.expectPVC(test.NewDefaultPVCWithLabel())
 			})
 		})
+		Context("with an existing PVC", func() {
+			JustBeforeEach(func() {
+				t.reconcileCryostatFully()
+			})
+			Context("that successfully updates", func() {
+				BeforeEach(func() {
+					existing := test.NewDefaultPVC()
+					// Add some labels and annotations to test merging
+					metav1.SetMetaDataLabel(&existing.ObjectMeta, "my", "other-label")
+					metav1.SetMetaDataLabel(&existing.ObjectMeta, "another", "label")
+					metav1.SetMetaDataAnnotation(&existing.ObjectMeta, "my/custom", "other-annotation")
+					metav1.SetMetaDataAnnotation(&existing.ObjectMeta, "another/custom", "annotation")
+					t.objs = append(t.objs, test.NewCryostatWithPVCSpec(), existing)
+				})
+				It("should update metadata and resource requests", func() {
+					expected := test.NewDefaultPVC()
+					metav1.SetMetaDataLabel(&expected.ObjectMeta, "my", "label")
+					metav1.SetMetaDataLabel(&expected.ObjectMeta, "another", "label")
+					metav1.SetMetaDataLabel(&expected.ObjectMeta, "app", "cryostat")
+					metav1.SetMetaDataAnnotation(&expected.ObjectMeta, "my/custom", "annotation")
+					metav1.SetMetaDataAnnotation(&expected.ObjectMeta, "another/custom", "annotation")
+					expected.Spec.Resources.Requests[corev1.ResourceStorage] = resource.MustParse("10Gi")
+					t.checkPVC(expected)
+				})
+			})
+			Context("that fails to update", func() {
+				var cr *operatorv1beta1.Cryostat
+				BeforeEach(func() {
+					cr = test.NewCryostat()
+					t.objs = append(t.objs, cr)
+				})
+				JustBeforeEach(func() {
+					// Replace client with one that fails to update the PVC
+					pvc := test.NewDefaultPVC()
+					invalidErr := kerrors.NewInvalid(schema.ParseGroupKind("PersistentVolumeClaim"), pvc.Name, field.ErrorList{
+						field.Forbidden(field.NewPath("spec"), "test error"),
+					})
+					t.Client = test.NewClientWithUpdateError(t.Client, pvc, invalidErr)
+					t.controller.Client = t.Client
+
+					// Update PVC spec in CR
+					err := t.Client.Get(context.Background(), types.NamespacedName{Name: "cryostat", Namespace: "default"}, cr)
+					Expect(err).ToNot(HaveOccurred())
+					cr.Spec.StorageOptions = test.NewCryostatWithPVCSpec().Spec.StorageOptions
+					err = t.Client.Update(context.Background(), cr)
+					Expect(err).ToNot(HaveOccurred())
+
+					// Expect an Invalid status error after reconciling
+					req := reconcile.Request{NamespacedName: types.NamespacedName{Name: "cryostat", Namespace: "default"}}
+					_, err = t.controller.Reconcile(context.Background(), req)
+					Expect(err).To(HaveOccurred())
+					Expect(kerrors.IsInvalid(err)).To(BeTrue())
+				})
+				It("should emit a PersistentVolumeClaimInvalid event", func() {
+					recorder := t.controller.EventRecorder.(*record.FakeRecorder)
```

Comment:
Could you explain this syntax here? Sorry, I couldn't understand it (first time seeing this).

Reply:
This is a [type assertion[(https://go.dev/ref/spec#Type_assertions). It's similar to a cast. In this case the controller's `EventRecorder` field is of the `record.EventRecorder` interface type. Instead of the real EventRecorder, which talks to the Kubernetes API server, the tests use the controller-runtime project's `FakeRecorder`. This also implements the `record.EventRecorder` interface, but only simulates talking to the API server. Because the test case needs to access the `Events` channel field of the `FakeRecorder` type, we use a type assertion to make our `recorder` variable of type `record.FakeRecorder` instead of the `record.EventRecorder` interface it implements.

Reply:
Oh right! Make sense! Thanks a lot for explaining this! I probably should go back to revise some Golang basics :D